### PR TITLE
S3C-1115: rf: removes JSON API usage in GCP

### PR DESCRIPTION
**Pull request by alexanderchan-scality**
Removes the JSON API usage in GCP

drop commit: using test Arsenal branch with changes made to lifecycle configuration model generator

Federation: https://github.com/scality/Federation/pull/1382
Integration: https://github.com/scality/Integration/pull/647

### CODE REVIEW

**Comment by alexanderchan-scality** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=ft/back-to-xml
SCALITY_FEDERATION_BRANCH=ft/back-to-xml
SCALITY_S3_BRANCH=ft/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by bennettbuchanan** [comment]
Looks like the Federation PR needs to be rebased on rel/7.4 to include this change: https://github.com/scality/Federation/pull/1368/files#diff-b106b10a40ec0b4affd4efa2884f0b9fR73.

File: constants.js [hunk 1]
```diff
@@ -70,6 +70,8 @@ const constants = {
     // https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html
     maximumAllowedPartCount: 10000,
 
+    gcpMaximumAllowdPartCount: 1024,
```
Comment (bennettbuchanan) [inline constants.js]: Typo: `gcpMaximumAllowdPartCount` => `gcpMaximumAllowedPartCount`

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

File: lib/data/external/GcpClient.js [hunk 2]
```diff
@@ -91,6 +95,198 @@ class GcpClient extends AwsClient {
...
+    createMPU(key, metaHeaders, bucketName, websiteRedirectHeader, contentType,
+        cacheControl, contentDisposition, contentEncoding, log, callback) {
+        const metaHeadersTrimmed = {};
+        Object.keys(metaHeaders).forEach(header => {
+            if (header.startsWith('x-amz-meta-')) {
+                const headerKey = header.substring(11);
+                metaHeadersTrimmed[headerKey] = metaHeaders[header];
+            }
+        });
+        Object.assign(metaHeaders, metaHeadersTrimmed);
```
Comment (bennettbuchanan) [inline lib/data/external/GcpClient.js]: Seems like we would not want to preserve the 'untrimmed' headers and store them with the 'trimmed' ones. Just want to confirm that this is intentional.

File: lib/data/wrapper.js [hunk 3]
```diff
@@ -696,6 +715,19 @@ const data = {
             });
         }
 
+        if (locationTypeMatchGCP && dataLocator.lenegth === 1) {
```
Comment (bennettbuchanan) [inline lib/data/wrapper.js]: Typo: `lenegth` => `length`

File: lib/data/external/GcpClient.js (hunk 2 above)
Comment (alexanderchan-scality) [reply lib/data/external/GcpClient.js]: this is definitely a bug. will fix this.

Comment (alexanderchan-scality) [reply lib/data/external/GcpClient.js]: this is definitely not intentional. will fix this

File: lib/data/wrapper.js (hunk 3 above)
Comment (alexanderchan-scality) [reply lib/data/wrapper.js]: will fix this.

File: constants.js (hunk 1 above)
Comment (alexanderchan-scality) [reply constants.js]: will fix this.

File: lib/data/external/GcpClient.js (hunk 2 above)
Comment (bennettbuchanan) [reply lib/data/external/GcpClient.js]: Thanks, @alexanderchan-scality. 🚀

File: lib/data/wrapper.js (hunk 3 above)
Comment (bennettbuchanan) [reply lib/data/wrapper.js]: Those `e` chars don't stay in one place! 😸

**Comment by alexanderchan-scality** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=ft/back-to-xml
SCALITY_FEDERATION_BRANCH=ft/back-to-xml
SCALITY_S3_BRANCH=ft/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=ft/back-to-xml
SCALITY_FEDERATION_BRANCH=ft/back-to-xml
SCALITY_S3_BRANCH=ft/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine clean_env

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

File: lib/data/external/GCP/GcpApis/completeMPU.js [hunk 4]
```diff
@@ -47,32 +46,17 @@ function completeMPU(params, callback) {
...
+        (numParts, next) => mpuHelper.composeFinal(numParts, params, next),
         (result, next) => mpuHelper.generateMpuResult(result, partList, next),
         (result, aggregateETag, next) =>
             mpuHelper.copyToMain(result, aggregateETag, params, next),
         (mpuResult, next) => {
             const delParams = {
-                Bucket: params.Bucket,
-                MPU: params.MPU,
-                Overflow: params.Overflow,
+                Bucket: params.MPU,
```
Comment (philipyoo) [inline lib/data/external/GCP/GcpApis/completeMPU.js]: why is `Bucket` using `params.MPU`?

File: lib/data/external/GCP/GcpApis/mpuHelper.js [hunk 5]
```diff
@@ -322,13 +265,13 @@ class MpuHelper {
             (metadata, next) => {
                 // copy the final object into the main bucket
                 const copyMetadata = Object.assign({}, metadata);
-                copyMetadata['scal-ETag'] = aggregateETag;
+                copyMetadata['scal-etag'] = aggregateETag;
```
[the remark this answers is not in the document]
Comment (alexanderchan-scality) [reply lib/data/external/GCP/GcpApis/mpuHelper.js]: Yes. I had went with all lower case to be consistent with all the other custom metadata headers.

File: lib/data/external/GCP/GcpApis/abortMPU.js [hunk 6]
```diff
@@ -8,15 +8,13 @@ const { logHelper } = require('../../utils');
...
- * @param {string} params.Overflow - overflow bucket name
  * @param {string} params.Key - object key
  * @param {number} params.UploadId - MPU upload id
  * @param {function} callback - callback function to call
  * @return {undefined}
  */
 function abortMPU(params, callback) {
-    if (!params || !params.Key || !params.UploadId ||
-        !params.Bucket || !params.MPU || !params.Overflow) {
+    if (!params || !params.Key || !params.UploadId || !params.Bucket) {
```
[the remark this answers is not in the document]
Comment (alexanderchan-scality) [reply lib/data/external/GCP/GcpApis/abortMPU.js]: there might be some flawed logic here. Will fix.

File: lib/data/external/GCP/GcpApis/completeMPU.js (hunk 4 above)
Comment (alexanderchan-scality) [reply lib/data/external/GCP/GcpApis/completeMPU.js]: same as above.

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Comment by rahulreddy** [comment]
@ironman-machine r=b756d7f3b571809dea06942c660d67276bb95684

**Comment by alexanderchan-scality** [comment]
@ironman-machine
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=rf/back-to-xml
SCALITY_FEDERATION_BRANCH=rf/back-to-xml
SCALITY_S3_BRANCH=rf/back-to-xml
try

**Pull request merged**